File: 2016/src/day17/task1/Task1.java
package day17.task1;

import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayDeque;

public class Task1 {
   Path path = Path.of("inputs", "day17", "input.txt");

   void task() {
      try {
         var passcode = Files.readString(this.path).trim();

         var queue = new ArrayDeque<State>();
         queue.add(new State("", 0, 0));

         var result = "";
         while (!queue.isEmpty()) {
            var state = queue.poll();
            var x = state.x();
            var y = state.y();
            if (x == 3 && y == 3) {
               result = state.path();
               break;
            }

            var doors = state.getUnlockedDoors(passcode);

            for (var i = 0; i < 4; i++) {
               if (!doors[i]) {
                  continue;
               }

               var nextState = switch (i) {
                  case Direction.up -> (y - 1 < 0) ? null : new State(state.path() + "U", x, y - 1);
                  case Direction.down -> (y + 1 >= 4) ? null : new State(state.path() + "D", x, y + 1);
                  case Direction.left -> (x - 1 < 0) ? null : new State(state.path() + "L", x - 1, y);
                  case Direction.right -> (x + 1 >= 4) ? null : new State(state.path() + "R", x + 1, y);
                  default -> {
                     throw new RuntimeException("Unreachable: " + i);
                  }
               };

               if (nextState == null) {
                  continue;
               }
               queue.add(nextState);
            }
         }
         System.out.println(result);
      } catch (Exception e) {
         e.printStackTrace();
         System.exit(1);
      }
   }

   public static void main(String[] args) {
      var start = Instant.now();

      new Task1().task();

      System.out.println("Done in " + Duration.between(start, Instant.now()).toString().substring(2).toLowerCase());
   }
}

record State(String path, int x, int y) {
   boolean[] getUnlockedDoors(String passcode) throws NoSuchAlgorithmException {
      var md5 = MessageDigest.getInstance("MD5");
      var hash = new BigInteger(1, md5.digest((passcode + this.path).getBytes())).toString(16);
      hash = "0".repeat(32 - hash.length()) + hash;

      var result = new boolean[4];
      for (var i = 0; i < 4; i++) {
         result[i] = switch (hash.charAt(i)) {
            case 'b', 'c', 'd', 'e', 'f' -> true;
            default -> false;
         };
      }

      return result;
   }
}

final class Direction {
   static final int up = 0;
   static final int down = 1;
   static final int left = 2;
   static final int right = 3;
}

File: 2017/src/day03/task1.ts
import path from "node:path";

const PATH = path.join("inputs", "day03", "input.txt");

async function task1() {
   const input = +(await Bun.file(PATH).text()).trim();

   const state = {
      value: 1,
      x: 0,
      y: 0,
      limit: 1,
      direction: "right",
   } satisfies State as State;

   while (state.value != input) {
      state.value++;
      switch (state.direction) {
         case "left":
            state.x--;
            if (state.x > state.limit) {
               break;
            }
            state.limit = -state.limit;
            state.direction = "down";
            break;
         case "right":
            state.x++;
            if (state.x < state.limit) {
               break;
            }
            state.limit = -state.limit;
            state.direction = "up";
            break;
         case "up":
            state.y--;
            if (state.y > state.limit) {
               break;
            }
            state.direction = "left";
            break;
         case "down":
            state.y++;
            if (state.y < state.limit) {
               break;
            }
            state.limit++;
            state.direction = "right";
            break;
         default:
            throw new Error(`${state.direction satisfies never as string} doesn't exist`);
      }
   }

   console.log(Math.abs(state.x) + Math.abs(state.y));
}

type State = {
   value: number;
   x: number;
   y: number;
   limit: number;
   direction: "up" | "right" | "down" | "left";
};

console.time("Task Done");
await task1();
console.timeEnd("Task Done");
